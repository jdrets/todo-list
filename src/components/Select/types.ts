export interface SelectTypes {
  placeholder?: string
  onChange: () => void
  label?: string
  value?: string | null
  fullWidth?: boolean
  name?: string
  required?: boolean
}

export interface SelectWrapperTypes {
  fullWidth?: boolean
}
