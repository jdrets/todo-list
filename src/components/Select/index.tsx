import React, { FunctionComponent } from 'react'
import { ChevronDown } from 'react-ionicons'

import {
  Wrapper,
  SelectField,
  Label,
  RequiredLabel,
  SelectFieldWrapper,
  IconWrapper
} from './styles'
import { SelectTypes } from './types'

const Select: FunctionComponent<SelectTypes> = ({
  placeholder,
  onChange,
  label,
  value,
  fullWidth,
  name,
  required,
  children
}) => (
  <Wrapper>
    <Label>
      {label} {required && <RequiredLabel>*</RequiredLabel>}
    </Label>
    <SelectFieldWrapper
      fullWidth={fullWidth}
    >
      <SelectField
        value={value || ''}
        placeholder={placeholder}
        onChange={onChange}
        name={name}
        required
      >
        {children}
      </SelectField>
      <IconWrapper>
        <ChevronDown />
      </IconWrapper>
    </SelectFieldWrapper>
  </Wrapper>
)

export default Select
