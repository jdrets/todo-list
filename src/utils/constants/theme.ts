const palette = {
  purpleDark: '#4B00BC',
  purple: '#6600FF',
  purpleMedium: '#8533FF',
  purpleLight: '#E3D0FF',
  green: '#59BC6F',
  greenLight: '#E6F5E2',
  red: '#EB5757',
  redLight: '#FCE4E4',
  yellow: '#DD9900',
  yellowLight: '#FEF1D8',
  white: '#FFFFFF',
  black: '#000000',
  gray: '#898989',
  grayLight: '#E0E0E0'
}

const theme = {
  colors: {
    primaryColor: palette.purple,
    primaryColorMedium: palette.purpleMedium,
    primaryColorLight: palette.purpleLight,
    primaryColorDark: palette.purpleDark,
    ...palette
  },
  layout: {
    breakpoints: {
      medium: '768px'
    },
    mobile: {
      width: '100%',
      padding: '0 24px'
    },
    desktop: {
      width: '1050px'
    }
  }
}

export default theme
