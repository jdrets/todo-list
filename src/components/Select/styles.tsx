import styled from 'styled-components'

import theme from '../../utils/constants/theme'
import { SelectWrapperTypes } from './types'

export const Wrapper = styled.div`
  display: block;
`

export const SelectFieldWrapper = styled.div<SelectWrapperTypes>`
  align-items: center;
  border-radius: 4px;
  border: 1px solid ${theme.colors.grayLight};
  display: flex;
  justify-content: space-between;
  position: relative;
  resize: none;
  transition: 0.3s;
  width: ${props => props.fullWidth ? '100%' : 'auto'};

  &:hover {
    border-color: ${theme.colors.primaryColorLight};
  }
  
  &:focus {
    border-color: ${theme.colors.primaryColor};
  }
`

export const IconWrapper = styled.div`
  position: absolute;
  right: 10px;
  top: 10px;
  z-index: 1;
  
  svg {
    color: ${theme.colors.grayLight};
    height: auto;
  }
`

export const SelectField = styled.select`
  appearance: none;
  background-color: transparent;
  border: none;
  outline: none;
  padding: 12px;
  position: relative;
  width: 100%;
  z-index: 2;
`

export const Label = styled.div`
  font-size: 14px;
  margin-bottom: 4px;
  font-weight: 300;
`

export const RequiredLabel = styled.span`
  color: ${theme.colors.red};
`
