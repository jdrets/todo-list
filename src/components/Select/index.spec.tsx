import React from 'react'
import { ThemeProvider } from 'styled-components'

import { render, fireEvent } from '@testing-library/react'

import Select from '.'
import theme from '../../utils/constants/theme'

const setup = (mockData = {}) => {
  return render(
    <ThemeProvider theme={theme}>
      <Select
        placeholder="dummy placeholder"
        onChange={jest.fn()}
        label="dummy label"
        value="dummy value"
        fullWidth
        name="dummy name"
        required
        {...mockData}
      >
        <option value="dummy value">dummy option</option>
      </Select>
    </ThemeProvider>
  )
}

describe('<Select />', () => {
  test('should match with snapshot', () => {
    const { container } = setup()

    expect(container).toMatchSnapshot()
  })

  test('should call onChange function when select is change', () => {
    const onChangeMock = jest.fn()
    const { queryByPlaceholderText } = setup({
      onChange: onChangeMock
    })
    const input = queryByPlaceholderText('dummy placeholder')

    fireEvent.change(input, {
      target: {
        value: 'new value'
      }
    })

    expect(onChangeMock).toHaveBeenCalled()
  })

  test('should render required label when required prop is true in select', () => {
    const { getByText } = setup({
      required: true
    })
    const requiredLabel = getByText('*')

    expect(requiredLabel).toBeInTheDocument()
  })

  test('should not render required labelasd when required prop is null in select', () => {
    const { queryByText } = setup({
      required: false
    })

    expect(queryByText('*')).toBeNull()
  })
})
