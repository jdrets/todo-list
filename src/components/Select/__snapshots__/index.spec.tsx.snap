// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Select /> should match with snapshot 1`] = `
<div>
  <div
    class="sc-gtssRu dICDYb"
  >
    <div
      class="sc-jSFkmK UNLJ"
    >
      dummy label
       
      <span
        class="sc-gKAblj hqUYfj"
      >
        *
      </span>
    </div>
    <div
      class="sc-dlnjPT dtLmxg"
    >
      <select
        class="sc-eCApGN daPEYn"
        name="dummy name"
        placeholder="dummy placeholder"
        required=""
      >
        <option
          value="dummy value"
        >
          dummy option
        </option>
      </select>
      <div
        class="sc-hKFyIo bVSfKK"
      >
        <span
          class="sc-bdnylx kpuqbH"
          color="#000000"
          height="22px"
          rotate="0"
          width="22px"
        >
          <svg
            class=""
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            
            <path
              d="M112 184l144 144 144-144"
              fill="none"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="48"
            />
          </svg>
        </span>
      </div>
    </div>
  </div>
</div>
`;
